### PR TITLE
Using Octokit for Github API requests

### PR DESCRIPTION
**Pull request by awaage**
This is hopefully one step in the right direction, to use Octokit for Github API requests.
This directly addresses: https://github.com/skywinder/github-changelog-generator/issues/222

I really would like to get this issue resolved: https://github.com/skywinder/github-changelog-generator/issues/361

For review purposes, I simply created a new fetcher (OctoFetcher). The follow-up to this PR should be to remove the old Fetcher. Alternatively, we could use factory pattern and support both, but I don't really see a point to it.

### CODE REVIEW

File: lib/github_changelog_generator/octo_fetcher.rb [hunk 1]
```diff
@@ -0,0 +1,293 @@
+module GitHubChangelogGenerator
+  # A Fetcher responsible for all requests to GitHub and all basic manipulation with related data
+  # (such as filtering, validating, e.t.c)
+  #
+  # Example:
+  # fetcher = GitHubChangelogGenerator::OctoFetcher.new options
+
+  class OctoFetcher
+    PER_PAGE_NUMBER   = 100
```
Comment (awaage) [inline lib/github_changelog_generator/octo_fetcher.rb]: Changed this to use the maximum PER_PAGE_NUMBER. This helps with large repositories, where we need to minimize API calls

File: lib/github_changelog_generator/octo_fetcher.rb [hunk 2]
```diff
@@ -0,0 +1,293 @@
+module GitHubChangelogGenerator
```
Comment (awaage) [inline lib/github_changelog_generator/octo_fetcher.rb]: I tried to maintain all the interfaces, and use similar method calls, logic flow to the original Fetcher class. A follow-up PR can be done to clean it up even more, but this hopefully aids in review.

File: lib/github_changelog_generator/generator/generator_fetcher.rb [hunk 3]
```diff
@@ -72,7 +72,8 @@ def set_date_from_event(event, issue)
       else
         begin
           commit = @fetcher.fetch_commit(event)
-          issue[:actual_date] = commit[:author][:date]
+          issue[:actual_date] = commit[:commit][:author][:date]
+          # issue[:actual_date] = commit[:author][:date]
         rescue
```
Comment (awaage) [inline lib/github_changelog_generator/generator/generator_fetcher.rb]: One difference in the way the commits are returned with Octokit. This was a nasty bug.

File: lib/github_changelog_generator/octo_fetcher.rb (hunk 1 above)
Comment (skywinder) [inline lib/github_changelog_generator/octo_fetcher.rb]: Perfect! 👍

**Comment by skywinder** [comment]
@awaage Thank you! It's really nice work.
Can you also cover new fetcher by test at least for most important functions as well as old fetcher?

According factory - I agree with you, we don't need to handle 2 fetchers. octokit should replace it completely. 

I prefer to work on this branch separately and merge to master, when old fetcher will completely eliminated.

**Comment by awaage** [comment]
I added specs for original fetcher. 
Using VCR gem

https://github.com/skywinder/github-changelog-generator/pull/381/commits/951a9c3a21ffa42ee3e9bdc7adfee642a91de095

**Comment by awaage** [comment]
Added specs for the new octo_fetcher 
Commit ( https://github.com/skywinder/github-changelog-generator/pull/381/commits/434cc850ea05db847625845b8b6b92f310f460e3 )

I standardized the OctoFetcher to return hashes/arrays with stringified keys, as opposed to the Sawyer::Resource type. This was to keep it simple, and as similar as possible to old code / tests.

I had to refactor some of the code so that it uses hash access with string keys rather than accessing as methods/objects.

**Comment by olleolleolle** [comment]
(@awaage, very excited about this! Super-cool work.)

**Comment by awaage** [comment]
@olleolleolle thanks! I ran into some API limits testing on my big repository last night. Hope to get to some real testing this weekend!

**Comment by awaage** [comment]
This PR is done

**Comment by skywinder** [comment]
@awaage Nice! Please, run `rubocop -a` inside the project and commit corrections (to make syntax unified and make Travis checks pass https://travis-ci.org/skywinder/github-changelog-generator/builds/132438761 )

**Comment by olleolleolle** [comment]
First thought: Can this change also remove the `github_api` gem as a dependency?

**Comment by awaage** [comment]
@olleolleolle - the follow up is to remove Fetcher completely, which means remove the specs and the dependencies. I can do that now, but I left it initially for comparison purpose.

**Comment by awaage** [comment]
I have created a new PR #384 to clean up, remove old fetcher, and make rubocop happy. Please see comments there for more information.

**Comment by olleolleolle** [comment]
I like the changes, and I'm glad you decided to go to these lengths to make the most of this change.

This puts us in a better position to improve debug logging or use any other Faraday tools that are available.

**Comment by olleolleolle** [comment]
@awaage Hello, again!

I got inspired to work further on this change, by extending your branch.

I'll make a Pull Request on top of `retentionscience:aw/octokit-fetcher`, so that you can review the changes in good time. 

Things I've done now: 

- drop github_api (you'd done a great job of preparing that)
- depend on multi_json directly (it was a dependency somewhere)
- upgrade colorize to 0.8.1 (to avoid deprecation warnings)
- use [VCR with its RSpec metadata](https://www.relishapp.com/vcr/vcr/v/3-0-1/docs/test-frameworks/usage-with-rspec-metadata) (add a `:vcr` just next to the `context`, before its `do` block declaration)

This had the effect that we got rid of:

- rack as a dependency (currently, Rack 2.x as a dependency stops Ruby < 2.2 from installing our gem)
- hashie as a dependency (currently working on stepping away from this in our specs, mostly deleting things)

Having fun!

**Comment by olleolleolle** [comment]
Duplicate of #422.  Closing this issue. Thank you!

**Pull request closed without merging**